File: clients/scala/src/main/scala/komus24/model/CityType.scala
package komus24.model

import komus24.util.StreamUtil

/**
 * TODO - Document
 */
sealed trait CityType {
    /**
     * Write CityType to output stream
     */
    def writeTo(stream: java.io.OutputStream): scala.Unit
}

object CityType {
    /**
     * TODO - Document
     *
     * @param size TODO - Document
     * @param blockSize TODO - Document
     * @param refills TODO - Document
     */
    case class Manhattan(size: komus24.model.Vec2Int, blockSize: komus24.model.Vec2Int, refills: Int) extends CityType {
        /**
         * Write Manhattan to output stream
         */
        override def writeTo(stream: java.io.OutputStream): scala.Unit = {
            StreamUtil.writeInt(stream, Manhattan.TAG)
            size.writeTo(stream)
            blockSize.writeTo(stream)
            StreamUtil.writeInt(stream, refills)
        }
    
        /**
         * Get string representation of Manhattan
         */
        override def toString(): String = {
            var stringBuilder = new StringBuilder("Manhattan { ")
            stringBuilder.append("size: ")
            stringBuilder.append(size)
            stringBuilder.append(", ")
            stringBuilder.append("blockSize: ")
            stringBuilder.append(blockSize)
            stringBuilder.append(", ")
            stringBuilder.append("refills: ")
            stringBuilder.append(refills)
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object Manhattan {
        val TAG: Int = 0
    
        /**
         * Read Manhattan from input stream
         */
        def readFrom(stream: java.io.InputStream): Manhattan = Manhattan(
            komus24.model.Vec2Int.readFrom(stream),
            komus24.model.Vec2Int.readFrom(stream),
            StreamUtil.readInt(stream)
        )
    }

    /**
     * TODO - Document
     *
     * @param cells TODO - Document
     */
    case class Inline(cells: Seq[String]) extends CityType {
        /**
         * Write Inline to output stream
         */
        override def writeTo(stream: java.io.OutputStream): scala.Unit = {
            StreamUtil.writeInt(stream, Inline.TAG)
            StreamUtil.writeInt(stream, cells.length)
            cells.foreach { value =>
                StreamUtil.writeString(stream, value)
            }
        }
    
        /**
         * Get string representation of Inline
         */
        override def toString(): String = {
            var stringBuilder = new StringBuilder("Inline { ")
            stringBuilder.append("cells: ")
            stringBuilder.append(cells)
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object Inline {
        val TAG: Int = 1
    
        /**
         * Read Inline from input stream
         */
        def readFrom(stream: java.io.InputStream): Inline = Inline(
            (0 until StreamUtil.readInt(stream)).map { _ =>
                StreamUtil.readString(stream)
            }
        )
    }

    /**
     * Read CityType from input stream
     */
    def readFrom(stream: java.io.InputStream): CityType = {
        StreamUtil.readInt(stream) match {
            case Manhattan.TAG => Manhattan.readFrom(stream)
            case Inline.TAG => Inline.readFrom(stream)
            case _ => throw new java.io.IOException("Unexpected tag value")
        }
    }
}

File: clients/java/src/main/java/komus24/model/CityType.java
package komus24.model;

import komus24.util.StreamUtil;

/**
 * City type
 */
public abstract class CityType {
    /**
     * Write CityType to output stream
     */
    public abstract void writeTo(java.io.OutputStream stream) throws java.io.IOException;

    /**
     * Read CityType from input stream
     */
    public static CityType readFrom(java.io.InputStream stream) throws java.io.IOException {
        switch (StreamUtil.readInt(stream)) {
            case Manhattan.TAG:
                return Manhattan.readFrom(stream);
            case Inline.TAG:
                return Inline.readFrom(stream);
            default:
                throw new java.io.IOException("Unexpected tag value");
        }
    }

    /**
     * Auto generated manhattan map
     */
    public static class Manhattan extends CityType {
        public static final int TAG = 0;
    
        /**
         * Map size
         */
        private komus24.model.Vec2Int size;
    
        /**
         * Map size
         */
        public komus24.model.Vec2Int getSize() {
            return size;
        }
    
        /**
         * Map size
         */
        public void setSize(komus24.model.Vec2Int value) {
            this.size = value;
        }
        /**
         * Size of a single block
         */
        private komus24.model.Vec2Int blockSize;
    
        /**
         * Size of a single block
         */
        public komus24.model.Vec2Int getBlockSize() {
            return blockSize;
        }
    
        /**
         * Size of a single block
         */
        public void setBlockSize(komus24.model.Vec2Int value) {
            this.blockSize = value;
        }
        /**
         * Number of refill stations
         */
        private int refills;
    
        /**
         * Number of refill stations
         */
        public int getRefills() {
            return refills;
        }
    
        /**
         * Number of refill stations
         */
        public void setRefills(int value) {
            this.refills = value;
        }
    
        public Manhattan(komus24.model.Vec2Int size, komus24.model.Vec2Int blockSize, int refills) {
            this.size = size;
            this.blockSize = blockSize;
            this.refills = refills;
        }
    
        /**
         * Read Manhattan from input stream
         */
        public static Manhattan readFrom(java.io.InputStream stream) throws java.io.IOException {
            komus24.model.Vec2Int size;
            size = komus24.model.Vec2Int.readFrom(stream);
            komus24.model.Vec2Int blockSize;
            blockSize = komus24.model.Vec2Int.readFrom(stream);
            int refills;
            refills = StreamUtil.readInt(stream);
            return new Manhattan(size, blockSize, refills);
        }
    
        /**
         * Write Manhattan to output stream
         */
        @Override
        public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
            StreamUtil.writeInt(stream, TAG);
            size.writeTo(stream);
            blockSize.writeTo(stream);
            StreamUtil.writeInt(stream, refills);
        }
    
        /**
         * Get string representation of Manhattan
         */
        @Override
        public String toString() {
            StringBuilder stringBuilder = new StringBuilder("Manhattan { ");
            stringBuilder.append("size: ");
            stringBuilder.append(String.valueOf(size));
            stringBuilder.append(", ");
            stringBuilder.append("blockSize: ");
            stringBuilder.append(String.valueOf(blockSize));
            stringBuilder.append(", ");
            stringBuilder.append("refills: ");
            stringBuilder.append(String.valueOf(refills));
            stringBuilder.append(" }");
            return stringBuilder.toString();
        }
    }

    /**
     * Fixed map
     */
    public static class Inline extends CityType {
        public static final int TAG = 1;
    
        /**
         * Each string represents a row in the city
         */
        private String[] cells;
    
        /**
         * Each string represents a row in the city
         */
        public String[] getCells() {
            return cells;
        }
    
        /**
         * Each string represents a row in the city
         */
        public void setCells(String[] value) {
            this.cells = value;
        }
    
        public Inline(String[] cells) {
            this.cells = cells;
        }
    
        /**
         * Read Inline from input stream
         */
        public static Inline readFrom(java.io.InputStream stream) throws java.io.IOException {
            String[] cells;
            cells = new String[StreamUtil.readInt(stream)];
            for (int cellsIndex = 0; cellsIndex < cells.length; cellsIndex++) {
                String cellsElement;
                cellsElement = StreamUtil.readString(stream);
                cells[cellsIndex] = cellsElement;
            }
            return new Inline(cells);
        }
    
        /**
         * Write Inline to output stream
         */
        @Override
        public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
            StreamUtil.writeInt(stream, TAG);
            StreamUtil.writeInt(stream, cells.length);
            for (String cellsElement : cells) {
                StreamUtil.writeString(stream, cellsElement);
            }
        }
    
        /**
         * Get string representation of Inline
         */
        @Override
        public String toString() {
            StringBuilder stringBuilder = new StringBuilder("Inline { ");
            stringBuilder.append("cells: ");
            stringBuilder.append("[ ");
            for (int cellsIndex = 0; cellsIndex < cells.length; cellsIndex++) {
                if (cellsIndex != 0) {
                    stringBuilder.append(", ");
                }
                String cellsElement = cells[cellsIndex];
                stringBuilder.append('"' + cellsElement + '"');
            }
            stringBuilder.append(" ]");
            stringBuilder.append(" }");
            return stringBuilder.toString();
        }
    }
}